cdap-app-templates/cdap-etl/cdap-etl-realtime-app/src/main/java/co/cask/cdap/app/etl/realtime/ETLWorker.java

```diff
@@ -227,8 +241,31 @@ public void run(DatasetContext context) throws Exception {
       // to be persisted in the sink.
       for (Object sourceData : sourceEmitter) {
         try {
-          for (Object object : transformExecutor.runOneIteration(sourceData)) {
-            dataToSink.add(object);
+          TransformResponse transformResponse = transformExecutor.runOneIteration(sourceData);
+          while (transformResponse.getEmittedRecords().hasNext()) {
+            dataToSink.add(transformResponse.getEmittedRecords().next());
+          }
+          final List<TransformResponse.TransformError> errorRecords = transformResponse.getErrorRecords();
+
+          for (final TransformResponse.TransformError response : errorRecords) {
+            final String datasetName = tranformIdToDatasetName.get(response.getTransformId());
+            getContext().execute(new TxRunnable() {
+              @Override
+              public void run(DatasetContext context) throws Exception {
+                Table errorTable = context.getDataset(datasetName);
+                String rowkeyPrefix = System.currentTimeMillis() + SEPARATOR + appName + SEPARATOR +
```

Comment:
row key can be composed a little better to save bytes.  Timestamp is a long, followed by a string for app name, separator, transformid, then UUID is 128 bits.  Bytes has a method to interact with UUID.